File: datasets/Part 10 - Model Selection & Boosting/Section 48 - Model Selection/1_K_FOLD_CROSS_VALIDATION.py
# -*- coding: utf-8 -*-
"""
Created on Tue Feb 14 22:19:02 2023

@author: Imanol
"""

# K - FOLD CROSS VALIDATION


# Cómo importar las librerías
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd


# Importar el data set
dataset = pd.read_csv("Social_Network_Ads.csv")

X = dataset.iloc[:,[2,3]].values
y = dataset.iloc[:,-1].values

# Dividir el data set en training y testing
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.25, random_state = 0) 

# Escalado de variables
from sklearn.preprocessing import StandardScaler
sc_X = StandardScaler()
X_train = sc_X.fit_transform(X_train)
X_test = sc_X.transform(X_test)

# Ajustar el SVM con kernel en el Conjunto de Entrenamiento
from sklearn.svm import SVC
classifier = SVC(kernel= "rbf",
                 random_state= 0)
classifier.fit(X_train, y_train)

# Predicción de los resultados con el Conjunto de Testing
y_pred = classifier.predict(X_test)

# Elaborar una matriz de confusión
from sklearn.metrics import confusion_matrix
cm = confusion_matrix(y_test, y_pred)

# Aplicar k-fold cross validation +++++++++++++++++++++++++++++++++++++++++++++++++++++++++
from sklearn.model_selection import cross_val_score
accuracies = cross_val_score(estimator= classifier, 
                             X= X_train, y= y_train,
                             cv= 10)
accuracies.mean()
accuracies.std()

File: datasets/Part 10 - Model Selection & Boosting/Section 48 - Model Selection/1_K_FOLD_CROSS_VALIDATION.R
# K-FOLD CROSS VALIDATION

# Importar el dataset
dataset <- read.csv('Social_Network_Ads.csv')
dataset <- dataset[, 3:5]


# Dividir los datos en training y testing
#install.packages("caTools")
library(caTools)
set.seed(123)
split <- sample.split(dataset$Purchased, SplitRatio = 0.75)
training_set = subset(dataset, split == TRUE)
testing_set = subset(dataset, split == FALSE)

# Escalado de variables
training_set[,1:2] = scale(training_set[,1:2])
testing_set[,1:2] = scale(testing_set[,1:2])


# Ajustar el clasificador en el Conjunto de Entrenamiento
library(e1071)
classifier = svm(formula = Purchased ~ .,
                 data = training_set,
                 type = "C-classification",
                 kernel = "radial")


# Predicción de los resulados con el Conjunto de Testing
y_pred = predict(classifier, newdata = testing_set[,-3])

# Crear la matriz de confusión
cm = table(testing_set[,3], y_pred)

# Aplicar el algoritmo de k-fold cross validation ++++++++++++++++++++++++++++++
library(caret)
folds = createFolds(training_set$Purchased, k = 10)
cv = lapply(folds, function(x) {
  training_fold = training_set[-x, ]
  test_fold = training_set[x, ]
  classifier = svm(formula = Purchased ~ .,
                   data = training_fold,
                   type = "C-classification",
                   kernel = "radial")
  y_pred = predict(classifier, newdata = test_fold[,-3])
  cm = table(test_fold[,3], y_pred)
  accuracy = (cm[1,1]+cm[2,2]) / (cm[1,1]+cm[1,2]+cm[2,1]+cm[2,2])
  return(accuracy)
})

accuracy = mean(as.numeric(cv))
accuracy_sd = sd(as.numeric(cv))